src/providers/MeetingProvider/MeetingManager.ts

```diff
@@ -214,29 +166,77 @@ export class MeetingManager implements AudioVideoObserver {
     this.audioVideoObservers = {};
   }
 
-  async join({
-    meetingInfo,
-    attendeeInfo,
-    deviceLabels = DeviceLabels.AudioAndVideo,
-    eventReporter,
-    meetingManagerConfig,
-  }: MeetingJoinData) {
-    if (meetingManagerConfig) {
-      this.meetingManagerConfig = meetingManagerConfig;
-    }
-    this.configuration = new MeetingSessionConfiguration(
-      meetingInfo,
-      attendeeInfo
+  async join(meetingJoinData: MeetingJoinData): Promise<void> {
+    const { meetingSessionConfiguration } = meetingJoinData;
+    const {
+      deviceLabels,
+      eventReporter,
+      logger,
+      enableWebAudio,
+      activeSpeakerPolicy,
+    } = this.setupMeetingManager(meetingJoinData, meetingSessionConfiguration);
+    this.meetingSessionConfiguration = meetingSessionConfiguration;
+    this.meetingId = this.meetingSessionConfiguration.meetingId;
+    this.logger = logger;
+
+    const deviceController = new DefaultDeviceController(logger, {
+      enableWebAudio: enableWebAudio,
+    });
+    this.meetingSession = new DefaultMeetingSession(
+      meetingSessionConfiguration,
+      logger,
+      deviceController,
+      eventReporter
     );
 
-    this.meetingRegion = meetingInfo.MediaRegion;
-    this.meetingId = this.configuration.meetingId;
+    this.audioVideo = this.meetingSession.audioVideo;
 
-    if (eventReporter) {
-      this.eventReporter = eventReporter;
-    }
+    // When an attendee leaves, we remove AudioVideoObservers and nullify AudioVideoFacade and MeetingSession object.
+    // This results into missing few meeting events triggered with audioVideoDidStop such as meetingEnded, meetingFailed
+    // and meetingStartFailed. We may also loose audioInputUnselected and videoInputUnselected events as we choose null
+    // devices when an attendee leaves. When a new AudioVideoFacade object is created remove and re-add the
+    // eventDidReceive observer which wont leak.
+    this.audioVideo.removeObserver(this.eventDidReceiveRef);
+    this.audioVideo.addObserver(this.eventDidReceiveRef);
 
-    await this.initializeMeetingSession(this.configuration, deviceLabels);
+    this.setupAudioVideoObservers();
+    this.setupDeviceLabelTrigger(deviceLabels);
+    await this.listAndSelectDevices(deviceLabels);
+    this.publishAudioVideo();
+    this.setupActiveSpeakerDetection(activeSpeakerPolicy);
+    this.meetingStatus = MeetingStatus.Loading;
+    this.publishMeetingStatus();
+  }
+
+  private setupMeetingManager(
+    meetingJoinData: MeetingJoinData,
+    meetingSessionConfiguration: MeetingSessionConfiguration
+  ): MeetingManagerJoinOptions {
+    const deviceLabels =
+      meetingJoinData?.options?.deviceLabels ?? DeviceLabels.AudioAndVideo;
```

Comment:
Can you once verify if this `??` operator is supported across browsers we support? Our official supported version for Chrome is 78+ and the MDN says this is supported 80 onwards.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator

Reply:
I would prefer use `||` instead.

Reply:
Seems like we are already using in the tooltip UI component: https://github.com/aws/amazon-chime-sdk-component-library-react/blob/main/src/components/ui/WithTooltip/Styled.tsx#L22.

Reply:
Noone uses ToolTip so I would not rely on that :)

Reply:
I have removed the usage of `??` and added a TODO for future.